tools/chain-util/pkg/export/export.go

```diff
@@ -0,0 +1,173 @@
+	}
+
+	return &ChainExporter{
+		bstore:  bstore,
+		dagserv: dserv,
+		Head:    headTS,
+		out:     out,
+	}, nil
+}
+
+// ChainExporter exports the blockchain contained in bstore to the out writer.
+type ChainExporter struct {
+	// Where the chain data is kept
+	bstore blockstore.Blockstore
+	// Makes traversal easier
+	dagserv format.DAGService
+	// the Head of the chain being exported
+	Head block.TipSet
+	// Where the chain data is exported to
+	out io.Writer
+}
+
+// Export will export a chain (all blocks and their messages) to the writer `out`.
+func (ce *ChainExporter) Export(ctx context.Context) error {
+	msgStore := chain.NewMessageStore(ce.bstore)
+	return chain.Export(ctx, ce.Head, ce, msgStore, ce, ce.out)
+}
+
+// LoadTipSet loads all the TipSet for a given TipSetKey.
+func (ce *ChainExporter) GetTipSet(key block.TipSetKey) (block.TipSet, error) {
+	var blks []*block.Block
+	for it := key.Iter(); !it.Complete(); it.Next() {
+		bsBlk, err := ce.bstore.Get(it.Value())
+		if err != nil {
+			return block.UndefTipSet, errors.Wrapf(err, "failed to load tipset %s", key.String())
+		}
+		blk, err := block.DecodeBlock(bsBlk.RawData())
+		if err != nil {
+			return block.UndefTipSet, errors.Wrapf(err, "failed to decode tipset %s", key.String())
+		}
+		blks = append(blks, blk)
+	}
+	return block.NewTipSet(blks...)
+}
+
+// ChainStateTree returns the state tree as a slice of IPLD nodes at the passed stateroot cid `c`.
+func (ce *ChainExporter) ChainStateTree(ctx context.Context, c cid.Cid) ([]format.Node, error) {
+	return newChainStateCollector(ce.dagserv).collectState(ctx, c)
+}
+
+func newChainStateCollector(dserv format.DAGService) *chainStateCollector {
+	return &chainStateCollector{
+		dagserv: dserv,
+	}
+}
+
+type chainStateCollector struct {
+	dagserv format.DAGService // Provides access to state tree.
+	state   []format.Node
+}
+
+// collectState recursively walks the state tree starting with `stateRoot` and returns it as a slice of IPLD nodes.
+// Calling this method does not have any side effects.
+func (csc *chainStateCollector) collectState(ctx context.Context, stateRoot cid.Cid) ([]format.Node, error) {
+	dagNd, err := csc.dagserv.Get(ctx, stateRoot)
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to load stateroot from dagservice %s", stateRoot)
+	}
+	csc.addState(dagNd)
+	seen := cid.NewSet()
+	for _, l := range dagNd.Links() {
+		if err := dag.Walk(ctx, csc.getLinks, l.Cid, seen.Visit); err != nil {
+			return nil, errors.Wrapf(err, "dag service failed walking stateroot %s", stateRoot)
+		}
+	}
+	return csc.state, nil
+
+}
+
+func (csc *chainStateCollector) getLinks(ctx context.Context, c cid.Cid) ([]*format.Link, error) {
+	nd, err := csc.dagserv.Get(ctx, c)
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to load link from dagservice %s", c)
+	}
+	csc.addState(nd)
+	return nd.Links(), nil
+
+}
+
+func (csc *chainStateCollector) addState(nd format.Node) {
+	csc.state = append(csc.state, nd)
+}
```

Comment:
Copy-Pasta from here
https://github.com/filecoin-project/go-filecoin/blob/4f71dca6ec0a18e3b19445c39ef06033e51a859e/internal/app/go-filecoin/plumbing/cst/chain_state.go#L233-L277

I'm not happy about this code duplication, but it seems better than exporting `newChainStateCollector` from chain_state.go, thoughts?

Reply:
Could you elaborate on why importing from chain_state.go is worse?  Without your context importing seems like the better choice.  You could put this type in the chain package or in a new `ipld`/`dag` or the util package if you don't want to export free functions from plumbing.

Reply:
Mainly because I didn't like the idea of exporting a free function from plumbing. I will move this into dag plumbing exposed as a `RecursiveGet` method.